### PR TITLE
Add tristate type for offload options (LP: #1956264)

### PR DESCRIPTION
**Pull request by nicolasbock**
Signed-off-by: Nicolas Bock <nicolas.bock@canonical.com>

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [x] New/changed keys in YAML format are documented.
- [x] \(Optional\) Adds example YAML for new feature.
- [x] \(Optional\) Closes an open bug in Launchpad. LP#1956264

### CODE REVIEW

**Comment by nicolasbock** [comment]
At this point it seems that some tests are not run. I will check and update this PR.

File: doc/netplan.md [hunk 1]
```diff
@@ -149,44 +149,44 @@ Virtual devices
 
 ``receive-checksum-offload`` (bool) – since **0.104**
 
-:    (networkd backend only) If set to true, the hardware offload for
-     checksumming of ingress network packets is enabled. When unset,
+:    (networkd backend only) If set to true (false), the hardware offload for
+     checksumming of ingress network packets is enabled (disable). When unset,
```
[the remark this answers is not in the document]
Comment (nicolasbock) [reply doc/netplan.md]: Fixed

**Comment by nicolasbock** [comment]
@rlaager Could I get another review? Thanks!

File: doc/netplan.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (rlaager) [reply doc/netplan.md]: This part looks good to me. I'm not a developer here, so I can't mark this as resolved.

**Comment by nicolasbock** [comment]
Thanks fort he feedback @rlaager ! :smile:

**Comment by nicolasbock** [comment]
Would it help you if I made a package available for you?

**Comment by rlaager** [comment]
> Would it help you if I made a package available for you?

That would speed it up a bit, yes, as that saves me from patching the package and also guilts me into doing the testing right away. ;) Ideally, I'm looking for whatever is in 20.04 + this patch. But if it's a git snapshot or something that's probably fine too. I've got spare systems in the right state for quick testing, so it's pretty easy to run the test.

**Comment by rlaager** [comment]
TL;DR: The changes work as intended. There is a pre-existing issue applying link settings that is only documented for `wakeonlan` that applies here too.

I tested today. I built the netplan from focal with this patch applied. Maybe that's not ideal from an upstream perspective (since I'm not building the commit on master), but that is what was easiest for me.

My netplan config has this (among other things):
```
  ethernets:
    enp4s0:
      dhcp4: no
      dhcp6: no
      generic-receive-offload: no
      large-receive-offload: no
    enp6s0:
      dhcp4: no
      dhcp6: no
      #generic-receive-offload: no
      #large-receive-offload: no
```

That seems to build the correct output:
```
# cat /run/systemd/network/10-netplan-enp4s0.link
[Match]
OriginalName=enp4s0

[Link]
WakeOnLan=off
GenericReceiveOffload=off
LargeReceiveOffload=off
```

It doesn't actually work:
```
# ethtool -k enp4s0 | grep receive-offload
generic-receive-offload: on
large-receive-offload: off
# ethtool -k enp6s0 | grep receive-offload
generic-receive-offload: on
large-receive-offload: off
```

It's not an issue of `off` vs `on`, as set `large-receive-offload` to `yes` doesn't change anything.

Here we can see why:
```
# SYSTEMD_LOG_LEVEL=debug udevadm test-builtin net_setup_link /sys/class/net/enp4s0
Trying to open "/etc/systemd/hwdb/hwdb.bin"...
Trying to open "/etc/udev/hwdb.bin"...
Trying to open "/usr/lib/systemd/hwdb/hwdb.bin"...
Trying to open "/lib/systemd/hwdb/hwdb.bin"...
Trying to open "/lib/udev/hwdb.bin"...
=== trie on-disk ===
tool version:          245
file size:         9954959 bytes
header size             80 bytes
strings            2164111 bytes
nodes              7790768 bytes
Load module index
Found container virtualization none.
timestamp of '/etc/systemd/network' changed
timestamp of '/run/systemd/network' changed
Parsed configuration file /usr/lib/systemd/network/99-default.link
Parsed configuration file /usr/lib/systemd/network/73-usb-net-by-mac.link
Parsed configuration file /run/systemd/network/10-netplan-enp4s0.link
Created link configuration context.
ID_NET_DRIVER=ixgbe
enp4s0: Device has name_assign_type=4
enp4s0: Config file /run/systemd/network/10-netplan-enp4s0.link matches device based on renamed interface name, ignoring
enp4s0: Config file /usr/lib/systemd/network/99-default.link is applied
enp4s0: Failed to get ACTION= property: No such file or directory
ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
enp4s0: Device has name_assign_type=4
Using default interface naming scheme 'v245'.
enp4s0: Policy *keep*: keeping existing userspace name
enp4s0: Device has addr_assign_type=3
enp4s0: MAC on the device already set by userspace
ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link
Unload module index
Unloaded link configuration context.
```

networkd refuses to apply 10-netplan-enp4s0.link because it is matching the renamed name. This is the same problem as documented in netplan for `wakeonlan`:
> Note: This will not work reliably for devices matched by name
only and rendered by networkd, due to interactions with device
renaming in udev. Match devices by MAC when setting wake on LAN.

That issue is not specific to `wakeonlan` but applies to anything that gets written to the `.link` files. It might be a good idea to document that differently. (The easy way would be to duplicate that into every one of these settings. There are probably other, better, ways.)

**Review by slyon — COMMENTED** [review]
Thank you very much for this improvement @nicolasbock and sorry for the late reply (I was out of office for a while).

Thanks for the testing @rlaager If I understand your comment correctly the code works as intended, but is affected by an udev/link matching issue, which can be worked around by matching on MAC address?!

The PR looks good to me overall! There is just one small spelling change (on/off -> true/false) that I'd like to push for; see inline comment. Furthermore, I'm wondering if this offload config is something that could be tested in an emulated (qemu/kvm) test environment and if so if it would make sense to add an integration test for it in `tests/integration/ethernets.py`? But it might need special offloading hardware (which is not available in qemu), right?

Also, running `make check` locally seems to fail for me, due to coverage issues, could you please rebase this branch to the latest `main` branch and double check that?

**Comment by rlaager** [comment]
> Thanks for the testing @rlaager If I understand your comment correctly the code works as intended

Correct.

>but is affected by an udev/link matching issue

Correct. The `.link` files that netplan writes out are completely ignored (in general, not specific to this setting inside it) because the interface has already been renamed by the point that `.link` file is read by networkd. networkd refuses to use the `.link` file because it is named by the _renamed_ (new) name of the interface.

Specifically, in my case, these are Intel NICs using the ixgbe driver. On early boot, they are named `eth0` and `eth1` and then renamed to `enp4s0` and `enp6s0` (in this particular machine, which has two single-port NICs; other machines have single two-port NICs and the same problem exists with slightly different new names).

`/run/systemd/network/10-netplan-enp4s0.network` is _not_ honored by networkd at all because it contains:

    [Match]
    Name=enp4s0

Now, I would probably argue that networkd should apply `.link` files like this based on the new name. But this is not happening by accident. The networkd code contains a message about exactly what is happening here. So it's clearly intentional on their part.

I'm pretty sure I tried matching on `eth1`, but that doesn't work either, because there is no `eth1` at that point. That left matching by MAC.

> which can be worked around by matching on MAC address?!

Correct. networkd will apply the `.link` rules that match on the MAC address even after the interface is renamed, because the MAC address matches.

Now, in my particular case, the interfaces are being put into a bridge, so the MAC addresses get changed once they are added into the bridge. So in that particular configuration, even with the MAC address matching, the `.link` files are only applied on boot. I can set settings (like generic-receive-offload) that show up in the `.link` file and have them apply on boot, but I can't change them (via netplan) on a running system without rebooting.

That's all suboptimal, but: 1) Setting this at boot is good enough for me. 2) This issue is about `.link` settings in general, not this specific new setting, so there's no need to hold up this PR on it.

This is clearly a known issue, as the netplan documentation for `wakeonlan` describes this exact issue, saying, "Note: This will not work reliably for devices matched by name only and rendered by networkd, due to interactions with device renaming in udev. Match devices by MAC when setting wake on LAN."

So my suggestion was to either: A) duplicate that not to these settings, or B) make it more general somehow.

> I'm wondering if this offload config is something that could be tested in an emulated (qemu/kvm) test environment and if so if it would make sense to add an integration test for it in tests/integration/ethernets.py?

Yes, this can be tested in a qemu/kvm VM. With either a virtio or e1000e NIC, `sudo ethtool -k INTERFACE_NAME | grep -v fixed` shows `generic-receive-offload` (among a few other things, but _not_ `large-receive-offload`). And it can be successfully set with `sudo ethtool -K INTERFACE_NAME gro off`.

File: src/networkd.c [hunk 2]
```diff
@@ -265,26 +265,33 @@ write_link_file(const NetplanNetDefinition* def, const char* rootdir, const char
         g_string_append_printf(s, "MTUBytes=%u\n", def->mtubytes);
 
     /* Offload options */
-    if (def->receive_checksum_offload)
-        g_string_append_printf(s, "ReceiveChecksumOffload=%u\n", def->receive_checksum_offload);
+    if (def->receive_checksum_offload_tristate != NETPLAN_TRISTATE_UNSET)
+        g_string_append_printf(s, "ReceiveChecksumOffload=%s\n",
+        (def->receive_checksum_offload_tristate ? "on" : "off"));
```
[the remark this answers is not in the document]
Comment (nicolasbock) [reply src/networkd.c:270]: Sure, fixed.

File: src/types.h [hunk 3]
```diff
@@ -346,6 +368,15 @@ struct netplan_net_definition {
...
     gboolean sriov_delay_virtual_functions_rebind;
+
+    /* Tristate offload options */
+    NetplanTristate receive_checksum_offload_tristate;
+    NetplanTristate transmit_checksum_offload_tristate;
+    NetplanTristate tcp_segmentation_offload_tristate;
+    NetplanTristate tcp6_segmentation_offload_tristate;
+    NetplanTristate generic_segmentation_offload_tristate;
+    NetplanTristate generic_receive_offload_tristate;
+    NetplanTristate large_receive_offload_tristate;
```
[the remark this answers is not in the document]
Comment (nicolasbock) [reply src/types.h:379]: Ok

**Comment by nicolasbock** [comment]
@slyon I addressed your inline comments. I am still working on the other issues.

File: src/types.h [hunk 4]
```diff
@@ -42,13 +51,324 @@ struct NetplanOptionalAddressType {
...
+        char *stable_id;
+        char *device;
+        GData* passthrough; /* See g_datalist* functions */
+    } nm;
+    struct NetplanNetworkdSettings {
+        char *unit;
+    } networkd;
+} NetplanBackendSettings;
+
+struct netplan_net_definition {
```
Comment (slyon) [inline src/types.h:195]: We might have a merge conflict here, as many of the definitions from `types.h` actually moved into `abi.h` recently.

Comment (nicolasbock) [reply src/types.h:195]: I had rebased this update and moved some of the changes into `abi.h`. I hope it was all correct :smile:

Comment (rlaager) [reply src/types.h:195]: No, definitely not. Look at types.h in this PR to see what I mean.

Comment (nicolasbock) [reply src/types.h]: Thanks @rlaager and @slyon . I see what you meant now. That rebase definitely didn't go as planned :/

**Comment by nicolasbock** [comment]
@slyon When trying to update the `focal` abi definitions (following the instructions in https://github.com/canonical/netplan/blob/main/abi-compat/README.md#howto-create-a-abi-reference) I am getting the following error:

```console
ubuntu@trusty-tamarin:~/pwd$ meson setup build --prefix=/usr
The Meson build system
Version: 0.53.2
Source dir: /home/ubuntu/pwd
Build dir: /home/ubuntu/pwd/build
Build type: native build

meson.build:1:0: ERROR: Meson version is 0.53.2 but project requires >= 0.61.0

A full log can be found at /home/ubuntu/pwd/build/meson-logs/meson-log.txt
```

Do I need to install `meson` from somewhere else?

**Comment by slyon** [comment]
> @slyon When trying to update the `focal` abi definitions (following the instructions in https://github.com/canonical/netplan/blob/main/abi-compat/README.md#howto-create-a-abi-reference) I am getting the following error:

Don't worry about that too much. I can deal with the ABI definitions at the end. Focal packages are too old, so we're still using `make` on Focal (and the github CI). Basically what you'd do is:
```
make clean && make
abidw libnetplan.so.0.0 --headers-dir include/ --header-file src/abi.h > abi-compat/focal_0.104.xml
```

But you also need a newer version of `abigail-tools` from https://launchpad.net/~slyon/+archive/ubuntu/ci-netplan – So don't worry about that. You can just leave out the changes to `abi-compat/*` if you want, and I can update it before we merge the PR.

File: src/abi.h [hunk 5]
```diff
@@ -149,6 +149,27 @@ typedef union {
...
+     * FALSE -> 0
+     * TRUE  -> 1
+     * UNSET -> 2
+     * 
+     * And the integer values can be used directly when
+     * converting to string.
+     */
+    NETPLAN_TRISTATE_FALSE, /* 0 */
+    NETPLAN_TRISTATE_TRUE,  /* 1 */
+    NETPLAN_TRISTATE_UNSET, /* 2 */
```
[the remark this answers is not in the document]
Comment (nicolasbock) [reply src/abi.h:170]: Yes, that looks good. Let me fix that.

File: doc/netplan.md [hunk 6]
```diff
@@ -79,6 +79,10 @@ Virtual devices
 
 ## Common properties for physical device types
 
+     **Note:** These options will not work reliably for devices matched by name
```
Comment (nicolasbock) [inline doc/netplan.md:82]: @rlaager and @slyon :

I moved this note a little up so that it generally refers to options rendered in the `.link` file.

**Comment by eli-schwartz** [comment]
> Do I need to install `meson` from somewhere else?

https://mesonbuild.com/Getting-meson.html

Despite that Ubuntu 20.04 is frozen at an old version of Meson, you can manually install the latest version (because its only dependency is python 3.7) and use that for personal development albeit obviously not to build Ubuntu packages.

**Comment by nicolasbock** [comment]
> > Do I need to install `meson` from somewhere else?
> 
> https://mesonbuild.com/Getting-meson.html
> 
> Despite that Ubuntu 20.04 is frozen at an old version of Meson, you can manually install the latest version (because its only dependency is python 3.7) and use that for personal development albeit obviously not to build Ubuntu packages.

Thanks for the information @eli-schwartz . I was just wondering within the context of the abi build instructions. Using `make` as @slyon suggested works great.

File: src/abi.h [hunk 7]
```diff
@@ -327,4 +348,13 @@ struct netplan_net_definition {
     /* netplan-feature: eswitch-mode */
     char* embedded_switch_mode;
     gboolean sriov_delay_virtual_functions_rebind;
+
```
[the remark this answers is not in the document]
Comment (slyon) [reply src/abi.h:351]: Nice! You're right. This is a special case where both size and content make sense using the new types as well as the old, and it does not shift the memory layout. In this case I guess we can just change the type of the previous fields.

**Comment by nicolasbock** [comment]
I am confused @slyon . Why revert the change to `WakeOnLan`? As you [commented](https://github.com/canonical/netplan/pull/270#discussion_r856301180) the `true`/`false` usage is more common.

File: doc/netplan.md (hunk 6 above)
Comment (slyon) [reply doc/netplan.md]: I made this comment a bit more specific, as it does not apply to all of the "Common properties for physical device types", but only the udev/.link changes. And actually I added some `udevadm test /sys/class/net/IFACE` command to `netplan apply`, which might resolve this issue in some cases. At least `netplan apply` is now able to apply the `*-offload` options at runtime (without reboot) during the integration tests. cc @rlaager

**Review by nicolasbock — COMMENTED** [review]
LGTM

**Review by slyon — APPROVED** [review]
Sorry, I took the liberty to push to your PR branch, in order to clean up a few things.

We do not need to squash the commits, as we'll do a squash-merge afterwards. Also, I've reverted the WakeOnLan= setting, because it is unrelated to this offloading PR and also the `off` value is special/explicit in this case and not your usual boolean (.c.f. https://systemd.network/systemd.link.html#WakeOnLan=)

As suggested by @rlaager the ABI compatibility is not actually broken by this change of `gboolean` to `NetplanTristate` as both use the size of an integer and are even value/content compatible, as NETPLAN_TRISTATE_UNSET defaults to the kernel value (TRUE). So we can avoid deprecating the old offloading fields. Thanks for this investigation!

Also, thanks for your example of using `ethtool` to get/set those values. I've used it to implement an integration tests, which will be run during `autopkgtest` inside a qemu/kvm VM and/or LXD container. I needed to add a small quirt to the CI workflow, in order to add `ethtool` as a test dependency. Truns out that those values (all but large-receive-offload) can be used on a `veth` device as well, which is what we're using inside the integration tests:
```
# ethtool -k eth42 | grep -v fixed
Features for eth42:
rx-checksumming: on
tx-checksumming: on
        tx-checksum-ip-generic: on
        tx-checksum-sctp: on
scatter-gather: on
        tx-scatter-gather: on
        tx-scatter-gather-fraglist: on
tcp-segmentation-offload: on
        tx-tcp-segmentation: on
        tx-tcp-ecn-segmentation: on
        tx-tcp-mangleid-segmentation: on
        tx-tcp6-segmentation: on
generic-segmentation-offload: on
generic-receive-offload: on
```

Finally, I've added a `udevadm test /sys/class/net/IFACE` command inside the `netplan apply` CLI, to make it actually apply those changes at runtime without a reboot. I think this migh solve the issue described above, at least in some cases (it works inside the integration test).

@nicolasbock Let me know what you think! From my POV this PR is now ready to be merged!
Thanks again for your continued contributions to the netplan project! And also thanks to all the other people involved in reviewing this PR!

**Comment by rlaager** [comment]
I backported this to 20.04 and retested. It works the same as before, as expected. So I'm happy with this.

@slyon I ended up figuring out the remaining piece for applying the changes in my scenario. I've taken that to a new PR #278.
